misk/src/main/kotlin/misk/client/HttpClientsConfig.kt

```diff
@@ -1,115 +1,75 @@
-import com.fasterxml.jackson.databind.type.TypeFactory
 import misk.config.Config
+import misk.logging.getLogger
 import misk.security.ssl.CertStoreConfig
 import misk.security.ssl.TrustStoreConfig
 import java.io.File
+import java.net.URL
 import java.time.Duration
 
-data class BackwardsCompatibleEndpointConfig(
-    //Common fields
-  val url: String? = null,
-  val envoy: HttpClientEnvoyConfig? = null,
-
-    //Legacy fields
-  val connectTimeout: Duration? = null,
-  val writeTimeout: Duration? = null,
-  val readTimeout: Duration? = null,
-  val pingInterval: Duration? = null,
-  val callTimeout: Duration? = null,
-  val maxRequests: Int = 128,
-  val maxRequestsPerHost: Int = 32,
-  val maxIdleConnections: Int = 100,
-  val keepAliveDuration: Duration = Duration.ofMinutes(5),
-  val ssl: HttpClientSSLConfig? = null,
-
-    //New fields
-  val clientConfig: HttpClientConfig? = null
-)
-
-class BackwardsCompatibleEndpointConfigConverter : com.fasterxml.jackson.databind.util.Converter<BackwardsCompatibleEndpointConfig, HttpClientEndpointConfig> {
-  override fun getInputType(typeFactory: TypeFactory) =
-      typeFactory.constructType(BackwardsCompatibleEndpointConfig::class.java)
-
-  override fun getOutputType(typeFactory: TypeFactory) =
-      typeFactory.constructType(HttpClientEndpointConfig::class.java)
-
-  override fun convert(value: BackwardsCompatibleEndpointConfig) =
-      //http://%s.%s.gns.square
-      if (value.clientConfig == null)
-        HttpClientEndpointConfig(
-            url = value.url,
-            envoy = value.envoy,
-            clientConfig = HttpClientConfig(
-                value.connectTimeout,
-                value.writeTimeout,
-                value.readTimeout,
-                value.pingInterval,
-                value.callTimeout,
-                value.maxRequests,
-                value.maxRequestsPerHost,
-                value.maxIdleConnections,
-                value.keepAliveDuration,
-                value.ssl
-            )
-        )
-      else
-        HttpClientEndpointConfig(
-            url = value.url,
-            envoy = value.envoy,
-            clientConfig = value.clientConfig
-        )
-}
-
+@JsonDeserialize(converter = BackwardsCompatibleClientsConfigConverter::class)
 data class HttpClientsConfig(
-  val defaultConnectTimeout: Duration? = null,
-  val defaultWriteTimeout: Duration? = null,
-  val defaultReadTimeout: Duration? = null,
-  val ssl: HttpClientSSLConfig? = null,
-  val defaultPingInterval: Duration? = null,
-  val defaultCallTimeout: Duration? = null,
-  @JsonDeserialize(contentConverter = BackwardsCompatibleEndpointConfigConverter::class)
+  @JsonAlias("hosts")
+  //Need to retain ordering, hence LinkedHashMap
+  val hostConfigs: LinkedHashMap<String, HttpClientConfig> = linkedMapOf(),
   val endpoints: Map<String, HttpClientEndpointConfig> = mapOf()
 ) : Config {
+  init {
+    validatePatterns()
+  }
+
+  private fun validatePatterns() = try {
```

Comment:
nit: put `validatePatterns` underneath `operator fun get`

Reply:
I explicitly did not want to put this into `get`, as I want the class to fail-fast on construction, instead of waiting until somebody actually requests a value with `get` and then gets an error.

Reply:
Agreed that makes sense. I mean underneath (declared after) not in to `operator fun get`